File: leetcode/merge_two_sorted_list/MergeTwoSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        // Note: The Solution object is instantiated only once and is reused by each test case.       
        ListNode * head = NULL; 
        ListNode * tail = NULL; 
        
        while (l1 != NULL && l2 != NULL) {
            if (l1->val < l2->val) {
                if (head == NULL) {
                    head = l1; 
                    tail = l1; 
                    l1 = l1->next; 
                    tail->next = NULL; 
                }
                else {
                    tail->next = l1; 
                    tail = tail->next; 
                    l1 = l1->next; 
                    tail->next = NULL; 
                }
            }else {
                if (head == NULL) {
                    head = l2; 
                    tail = l2; 
                    l2 = l2->next; 
                    tail->next = NULL; 
                }
                else {
                    tail->next = l2; 
                    tail = tail->next; 
                    l2 = l2->next; 
                    tail->next = NULL; 
                }
            }
        }
        
        if (l1 == NULL && l2 != NULL) {
            if (head == NULL) {
                return l2; 
            }
            else {
                tail->next = l2; 
            }
        }
        else if (l1 != NULL && l2 == NULL){
            if (head == NULL) {
                return l1; 
            }
            else {
                tail->next = l1; 
            }
        }
        
        return head; 
    }
};

File: leetcode6/merge_two_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 * 
 * Merge two sorted linked list. 
 * Two pointers iterating through two lists, and add the node with smaller value. 
 * Make sure the discuss the case whether the result list is empty or not. After 
 * exhaust either list, need to check whether the other list is empty or not. Append 
 * the other list to the end of result list if not empty. 
 */
public class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode head = null; 
        ListNode tail = null; 
        
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                if (tail == null) {
                    head = l1; 
                    tail = l1; 
                    l1 = l1.next; 
                    tail.next = null; 
                } else {
                    tail.next = l1;
                    l1 = l1.next; 
                    tail = tail.next; 
                    tail.next = null; 
                }
            } else {
                if (tail == null) {
                    head = l2; 
                    tail = l2; 
                    l2 = l2.next; 
                    tail.next = null; 
                } else {
                    tail.next = l2; 
                    l2 = l2.next; 
                    tail = tail.next; 
                    tail.next = null; 
                }
            }
        }
        
        if (l1 != null) {
            if (tail == null) {
                head = l1; 
            } else {
                tail.next = l1; 
            }
        } else if (l2 != null){
            if (tail == null) {
                head = l2; 
            } else {
                tail.next = l2; 
            }
        }
        
        return head; 
    }
}
